### PR TITLE
fix: Replace coroutines with tasks to avoid RuntimeError

### PR DESCRIPTION
**Pull request by dav-pascual**
Fixing code block in `_opentack_gather_responses` method to avoid RuntimeError when a coroutine is reused in the loop.

Including a unit test for this method as part of the commit.

The issue was that a coroutine (aka the async function calls parameters) cannot be reused with await or gather, hence when a ServerError exception was raised, it caused a RunTime error. A workarround for this is awaiting using tasks instead of couroutines directly, and recreate the tasks if necessary to call the coroutines again.

### CODE REVIEW

**Comment by pvoborni** [comment]
I think this patch doesn't solve the issue. The bug seems to be introduced in https://github.com/neoave/mrack/commit/8ef831297a99c47eb270a6ea4f5e33ed20c8caee The regression is that on re-try the functions are not called again. 

The whole idea is that on server error we call openstack again to get the data to try to avoid the ServerError.

**Comment by dav-pascual** [comment]
> I think this patch doesn't solve the issue. The bug seems to be introduced in [8ef8312](https://github.com/neoave/mrack/commit/8ef831297a99c47eb270a6ea4f5e33ed20c8caee) The regression is that on re-try the functions are not called again.
> 
> The whole idea is that on server error we call openstack again to get the data to try to avoid the ServerError.

@pvoborni 
I will review this further with @Tiboris 
but, what was preventing to re-call the functions was the incorrect use of `asyncio.gather` which cannot be reused with coroutines, hence provoking the RunTime error after the first failed iteration of the while loop and preventing the retry.

**Comment by pvoborni** [comment]
I don't think I was referring to that.

This method is called in e.g.:
```python
_, _, self.limits, _, _ = await self._openstack_gather_responses(
            self._load_flavors(),
            self._load_images(image_names),
            self.nova.limits.show(),
            self._load_networks(),
            self._load_ip_availabilities(),
        )
```
On server error we need to run again the methods. Resp. it would be good to re-run only the failed one., e.g. `self._load_images(image_names)`.  Wrapping a coroutine with task will not accomplish the re-run of the function call which created the coroutine. We probably need to pass the functions and their params to `_openstack_gather_responses` so that inside it can be re-run again. 

We can also say that the created test is not testing the scenario correctly. There is no re-run of initial function. Proper test would have e.g. 2 calls of `coro2` inside the `_openstack_gather_responses` where one would end with `ServerError` and second would succeed thus the function would return `result[2]` with proper result.

File: tests/unit/test_openstack.py [hunk 1]
```diff
@@ -594,3 +601,44 @@ async def test_network_picking_no_spread(self, x):
...
+        provider = OpenStackProvider()
+
+        # Test successful outcome
+        result = await provider._openstack_gather_responses([coro1(), [], {}])
+        assert result[0] == "result1"
+
+        # Test that we get successful outcome after getting a ServerError
+        error_flag = True
+        result = await provider._openstack_gather_responses(
+            [coro1(), [], {}], [coro2(), [], {}]
```
Comment (Tiboris) [inline tests/unit/test_openstack.py:635]: I think you forgot brackets here
```suggestion
            [coro1, [], {}], [coro2, [], {}]
```

**Review by pvoborni — APPROVED** [review]
Looks good to me and provisioning worked.

File: tests/unit/test_openstack.py (hunk 1 above)
Comment (dav-pascual) [reply tests/unit/test_openstack.py]: thanks, fixed

**Review by Tiboris — APPROVED** [review]
Thanks @dav-pascual for the patch and @pvoborni for knowledge sharing session.

**Pull request merged**